accelerate-llvm-ptx/src/Data/Array/Accelerate/LLVM/PTX/CodeGen/Base.hs

```diff
@@ -325,6 +328,101 @@ atomicAdd_f t addr val =
+      IntegralNumType TypeWord8  -> cast_shfl_cast_i
+      IntegralNumType TypeWord16 -> cast_shfl_cast_i
+      IntegralNumType TypeWord32 -> cast_shfl_cast_i
+
+      FloatingNumType TypeHalf  -> cast_shfl_cast_f
+      FloatingNumType TypeFloat -> shfl_down_float a delta -- no cast needed
+
+      -- TODO: split 64-bit primitives into two 32-bit integers somehow
+      IntegralNumType TypeInt64  -> undefined
+      IntegralNumType TypeWord64 -> undefined
+      FloatingNumType TypeDouble -> undefined
+    VectorScalarType _ -> error "is this a thing? VectorScalarType"
+  where
+    cast_shfl_cast_i :: (IsIntegral a, IsNum a) => CodeGen PTX (Operands a)
+    cast_shfl_cast_i = A.fromIntegral integralType numType a >>= \a' -> shfl_down_i32 a' delta >>= A.fromIntegral integralType numType
+
+    cast_shfl_cast_f :: (IsFloating a, IsNum a) => CodeGen PTX (Operands a)
+    cast_shfl_cast_f = A.toFloating numType floatingType a >>= \a' -> shfl_down_float a' delta >>= A.toFloating numType floatingType
+
+
+shfl_up_i32 :: Operands Int32                -- give up
+            -> Operands Word32               -- offset
+            -> CodeGen PTX (Operands Int32) -- received from down
+shfl_up_i32   = mk_shfl "up" "i32" (ScalarPrimType scalarType)
+
+shfl_up_float :: Operands Float                -- give up
+              -> Operands Word32               -- offset
+              -> CodeGen PTX (Operands Float) -- received from down
+shfl_up_float = mk_shfl "up" "f32" (ScalarPrimType scalarType)
+
+shfl_down_i32 :: Operands Int32                -- give up
+              -> Operands Word32               -- offset
+              -> CodeGen PTX (Operands Int32) -- received from down
+shfl_down_i32   = mk_shfl "down" "i32" (ScalarPrimType scalarType)
+
+shfl_down_float :: Operands Float                -- give up
+                -> Operands Word32               -- offset
+                -> CodeGen PTX (Operands Float) -- received from down
+shfl_down_float = mk_shfl "down" "f32" (ScalarPrimType scalarType)
+
+
+-- Only works for 32-bit int or floats!
+mk_shfl :: (IsPrim a)
+        => Label -- direction ("up","down","bfly","idx") (the latter two probably represent the "xor" and "get from an index" variants)
+        -> Label -- the type, "i32" or "f32"
+        -> PrimType a
+        -> Operands a                  -- value to give
+        -> Operands Word32             -- delta
+        -> CodeGen PTX (Operands a)   -- value received
+mk_shfl mode typ pt val delta = call (
+  -- starting CUDA 9.0, the normal `shfl` primitives are removed in favour of the newer `shfl_sync` ones:
+  -- they behave the same, except they start with a 'mask' argument specifying which threads participate in the shuffle.
+  -- Arguably, it'd be better to branch on the minimum requirements for `shfl_sync`, or maybe even to branch
+  -- (in real Haskell code) on the compute version of the gpu, but I couldn't find exact version numbers for these.
```

Comment:
`Foreign.CUDA.Driver.Utils.libraryVersion`

CUDA-9.0 corresponds to 9000.

Reply:
Is that 'better' than this way? The comment was more a reminder that branching on CUDA-9.0 is maybe not even ideal.

Reply:
yes, because `libraryVersion` is embedded from the `CUDA_VERSION` macro from `cuda.h`. The `cuda` package itself retains compatibility to (much) older versions, so cuda-0.9.0.0 might have been built against CUDA-8 (or earlier), for example.